lib/rust/parser/cdylib/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+///     // Allocate a new parser. If parsing more than one input, it would be more efficient to
+///     // allocate it once and reuse it.
+///     Parser *const parser = parser_new();
+///     // Allocate a parse result. If parsing more than one input, there are two efficient
+///     // approaches, depending on use pattern:
+///     // - Finish with each result before parsing the next (by using the resulting data, or
+///     //   copying it to memory used elsewhere); reuse one `ParseResult` for each input parsed.
+///     // - Keep `ParseResult`s to avoid needing to copy their contents; allocate new instances
+///     //   to store results of parsing additional inputs.
+///     ParseResult *const result = parse_result_alloc();
+///     // Parse the data.
+///     parser_run(input, input_len, result);
+///     // Gather the results.
+///     char *const sanitized_input = parse_result_get_input_data(result);
+///     const uint8_t *const output = parse_result_get_output_data(result);
+///     const uintptr_t output_len = parse_result_get_output_data_len(result);
+///     // Do something with the results.
+///     handle_parse_result(sanitized_input || input, output, output_len);
+///     // Free resources.
+///     parse_result_free(result);
+///     parser_free(parser);
+/// }
+/// ```
+#[derive(Debug, Default)]
+#[allow(missing_copy_implementations)] // Expected to contain non-Copy data in the future.
+pub struct Parser {
+    // The implementation doesn't currently support reusing parser resources, but when we add that
+    // optimization, the API supports it.
+}
+
+
+// === Implementation ===
+
+impl Parser {
+    fn run(&mut self, input: &[u8]) -> ParseResult {
+        let input = String::from_utf8_lossy(input);
+        let tree = enso_parser::Parser::new().run(&input);
+        let tree = match enso_parser::serialization::serialize_tree(&tree) {
+            Ok(tree) => tree,
+            // `Tree` does not contain any types with fallible `serialize` implementations, so this
+            // cannot fail.
+            Err(_) => default(),
+        };
+        let input = match input {
+            Cow::Owned(input) => Some(input),
+            _ => None,
+        };
+        ParseResult { input, tree }
+    }
+}
+
+
+// === C Interface ===
+
+/// Create a new parser. This operation is infallible; the result will never be NULL.
+#[no_mangle]
+pub extern "C" fn parser_new() -> *mut Parser {
+    Box::into_raw(default())
+}
+
+/// Parse the given source code, provided in the input parameters. The result will be written to the
+/// object provided in the output parameter; any previous state of the output object will be
+/// overwritten.
+///
+/// # Safety
+///
+/// The `parser` argument must be a value returned by `parser_new` that has never been passed to
+/// `parser_free`.
+/// The `input` argument must be a non-NULL pointer to a sequence of at least `input_len` readable
+/// bytes.
```

Comment:
What is the input encoding? UTF-8? E.g. `intput` consists of bytes of UTF-8 characters?